### PR TITLE
WIP feat: Ruby configuration format (alternative to YAML)

### LINKED ISSUE #367: feat: Ruby configuration format (alternative to YAML)
**May 2022 Update:** [See Below](https://github.com/bridgetownrb/bridgetown/issues/367#issuecomment-1124516677)

----

~~As an alternative (not a replacement!)~~ to Bridgetown's YAML format for the configuration file, we could offer a Ruby-based configuration file (`bridgetown.config.rb`). It might be something as simple as:

```ruby
Bridgetown.configure do |config|
  config.url = "https://mydomain.com"
  config.timezone = "UTC"

  config.pagination = {
    enabled: true,
    per_page: 30
  }

  config.environment :production do
    config.some_analytics_key = "UA-12345678"
  end
end
```

Feedback and further thoughts encouraged!

### PR DESCRIPTION
**Pull request by JuanVqz**
This is a 🙋 feature or enhancement.

## Summary

Thank you for your guidance in the discord channel.

I managed to don't break the `test_configuration.rb` file.
Now, I need to do the `bridgetown.config.rb` and see how we will read it, right?

## Context
Fixes #367

### CODE REVIEW

File: bridgetown-core/lib/bridgetown-core.rb [hunk 1]
```diff
@@ -250,6 +250,12 @@ def sanitized_path(base_directory, questionable_path)
       end
     end
 
+    def configure
+      s = Struct.new(:foo).new
+      yield s
+      puts s
+    end
```
Comment (JuanVqz) [inline bridgetown-core/lib/bridgetown-core.rb:257]:
```ruby
Bridgetown.configure do |config|
  config.foo = "Bar"
end
#<struct foo="Bar">
```
Now I can receive args in my `Bridgetown.configure` I'm wondering how can I receive `n` args without naming it.

**Comment by JuanVqz** [comment]
I think, I'm on the way ♦

```ruby
module Bridgetown
  extend self

  def configure
    yield configuration
  end

  def configuration
    @configuration ||= Configuration.new
  end
end

class Configuration
  attr_accessor :name, :site

  def initialize
    @name = name
    @site = site
  end
end

Bridgetown.configure do |config|
  config.name = "My name"
  config.site = {
    url: "htttps://foo.dev",
    name: "Foo Blog"
  }
end

puts Bridgetown.configuration.inspect
#<Configuration:0x0000560a853f2b28 @name="My name", @site={:url=>"htttps://foo.dev", :name=>"Foo Blog"}>

```

**Comment by jaredcwhite** [comment]
@JuanVqz BTW, if `Configuration` object were simply an instance of `HashWithDotAccess::Hash`, you could make it even simpler and not need to manually add tons of `attr_accessor`s.

File: bridgetown-core/lib/bridgetown-core/configuration.rb [hunk 2]
```diff
@@ -148,6 +148,8 @@ def safe_load_file(filename)
         Tomlrb.load_file(filename)
       when %r!\.ya?ml!i
         YAMLParser.load_file(filename) || {}
+      when %r!\.rb!i
+        Bridgetown.load_configure
```
Comment (JuanVqz) [inline bridgetown-core/lib/bridgetown-core/configuration.rb:152]: Should I name it `load_file` for consistency with the other two options?
```suggestion
        Bridgetown.load_file
```

Comment (jaredcwhite) [reply bridgetown-core/lib/bridgetown-core/configuration.rb:152]: Maybe `load_configuration_file`?

Comment (JuanVqz) [reply bridgetown-core/lib/bridgetown-core/configuration.rb:152]: changed, I tried to write a test to confirm Bridgetown is reading the `bridgetown.config.rb` file but I wasn't able to get it working, would you give me a  clue what should I do, please?

**Comment by jaredcwhite** [comment]
@JuanVqz I took a more closer look and have a couple of ideas how to finish this up…will get back to you next week.

**Comment by jaredcwhite** [comment]
Hey @JuanVqz, sorry for the delay…given the large volume of other ecosystem changes going on right with v1.0, I think I'm going to put this particular feature on hold for the time being. I'll leave it open for sure, but we'll probably revisit after our initial 1.0 release.

**Comment by JuanVqz** [comment]
> Hey @JuanVqz, sorry for the delay…given the large volume of other ecosystem changes going on right with v1.0, I think I'm going to put this particular feature on hold for the time being. I'll leave it open for sure, but we'll probably revisit after our initial 1.0 release.

that's ok, I'll be around.

**Comment by jaredcwhite** [comment]
I appreciate the work you put into this @JuanVqz…I've been starting to muse on a possible direction forward which is a bit different, so will be closing this PR. Feel free to add suggestions or ideas back over on the issue. https://github.com/bridgetownrb/bridgetown/issues/367#issuecomment-1124516677

**Pull request closed without merging**